pypga/modules/migen/spimachine.py

```diff
@@ -1,250 +1,328 @@
-from migen import *
-from migen.genlib.fsm import FSM, NextState
-from misoc.interconnect.csr import *
-from misoc.interconnect.stream import *
-from misoc.interconnect import wishbone
+from migen import Module, Signal, If, Cat, Mux
+from migen.genlib.fsm import FSM, NextState, NextValue
+from pypga.core.common import get_length
 
 
 class ClockGen(Module):
-    def __init__(self, width):
-        # Cycle duration - 1
-        self.div = Signal(width)
-        # If `extend` is asserted, the next half cycle is
-        # extended by the LSB of `div`
-        self.extend = Signal()
-        # Half cycle done
+    """
+    Clock signal generator that controls the length of one half cycle
+    of class SPIMachine.
+    """
+    def __init__(self, div):
+        # High, if half cycle is done
         self.done = Signal()
-        # Continue counting or re-load counter
-        self.count = Signal()
-
+        # Set to high to start counting
+        self.start = Signal()
+        # Current counter value
+        self.counter = Signal(get_length(div))
+        
         ###
+        
+        # Set done to True, when counter is zero, but only if the clock
+        # is not being started.
+        self.comb += [self.done.eq((self.counter == 0) & ~self.start)]
+    
+        fsm = FSM("IDLE")
```

Comment:
Now also implemented using a state machine, which makes it much more readable in my opinion.